Release GIL in PG destructor

tushar00jain opened the pull request:
Summary: Gloo PG doesn't release GIL, which results in python code hanging until the destructor completes. The destructor waits for all work on the PG to complete which can take a long time.

Test Plan: Ran

```
$ pytest --log-cli-level=INFO -vs torchft/local_sgd_integ_test.py
```

with a large timeout on the async work. Call to `gil_scoped_release` doesn't show up in the gdb stack trace.



cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k

d4l3k commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -2010,10 +2010,8 @@ communication mechanism.
-      py::class_<
-          ::c10d::ProcessGroup,
-          c10::intrusive_ptr<::c10d::ProcessGroup>,
-          ::c10d::PyProcessGroup>(module, "ProcessGroup",
```
We need to keep the ::c10d::PyProcessGroup which is a trampoline so you can use Python classes as ProcessGroups in C++.

Might need to change the `intrusive_ptr_no_gil_destructor_class_` signature or add a new template variant to support the 3 arg version

d4l3k reviewed (approved):
LGTM

tushar00jain commented:
@pytorchbot merge

tushar00jain committed:
Release GIL in PG destructor

Summary: Gloo PG doesn't release GIL, which results in python code hanging until the destructor completes. The destructor waits for all work on the PG to complete which can take a long time.

Test Plan: Ran

```
$ pytest --log-cli-level=INFO -vs torchft/local_sgd_integ_test.py
```

with a large timeout on the async work. Call to `gil_scoped_release` doesn't show up in the gdb stack trace.

Reviewers:

Subscribers:

Tasks:

Tags:

tushar00jain commented:
@pytorchbot merge
